packages/beacon-node/src/chain/chain.ts

```diff
@@ -853,6 +853,19 @@ export class BeaconChain implements IBeaconChain {
     if (headState) {
       this.opPool.pruneAll(headState, finalizedState);
     }
+
+    // Populate finalized pubkey cache
+    if (finalizedState?.epochCtx.isAfterEIP6110()) {
+      const pivotValidatorIndex = finalizedState.validators.length;
+      // TODO 6110: If we are not considering EIP-6914 see if there is any 
+      // data structure like OrderedMap in immutabe-js so we can do slicing instead of filter
+      const newFinalizedValidators = finalizedState.epochCtx.unfinalizedPubkey2index.filter((index, _pubkey) => index < pivotValidatorIndex);
+
+      newFinalizedValidators.forEach((index, pubkey) => {
+        finalizedState.epochCtx.addFinalizedPubkey(index, pubkey);
```

Comment:
this will only update unfinalized cache of the new finalized state or head state, other unfinalized cache of `CachedBeaconStateAllForks` objects are not updated, is it an issue? If not, need to drop some comments here

other option would be not to update unfinalized cache inside `addFinalizedPubkey` and have a separate method `deleteUnfinalizedPubkeys(pubkey: PubkeyHex[])`, then loop through all cached states and call  `deleteUnfinalizedPubkeys(pubkeys)`. The good thing of this is we can do all mutations in batch to improve performance

Reply:
> other option would be not to update unfinalized cache inside addFinalizedPubkey and have a separate method deleteUnfinalizedPubkeys(pubkey: PubkeyHex[]), then loop through all cached states and call deleteUnfinalizedPubkeys(pubkeys). The good thing of this is we can do all mutations in batch to improve performance

That's a good idea. Thanks